8265518: C1: Intrinsic support for Preconditions.checkIndex

kelthuzadx opened the pull request:
The JDK codebase re-created many variants of checkIndex(`grep -I -r 'cehckIndex' jdk/`). A notable variant is java.nio.Buffer.checkIndex, which annotated with @IntrinsicCandidate and it only has a corresponding C1 intrinsic version.

In fact, there is an utility method `jdk.internal.util.Preconditions.checkIndex`(wrapped by java.lang.Objects.checkIndex) that behaves the same as these variants of checkIndex, we can replace these re-created variants of checkIndex by Objects.checkIndex, it would significantly reduce duplicated code and enjoys performance improvement because Preconditions.checkIndex is @IntrinsicCandidate and it has a corresponding intrinsic method in HotSpot.

But, the problem is currently HotSpot only implements the C2 version of Preconditions.checkIndex. To reuse it global-widely in JDK code, I think we can firstly implement its C1 counterpart. There are also a few kinds of stuff we can do later:

1. Replace all variants of checkIndex by Objects.checkIndex in the whole JDK codebase.
2. Remove Buffer.checkIndex and obsolete/deprecate InlineNIOCheckIndex flag

Testing: cds, compiler and jdk

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265518](https://bugs.openjdk.java.net/browse/JDK-8265518): C1: Intrinsic support for Preconditions.checkIndex


### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to 954abc6e1bd086f4d0f8b2a9c717936b2b7604d3
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**) ⚠️ Review applies to 307d7a104eb1e02fba2c56b7243c6bb45f279f0b


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3615/head:pull/3615` \
`$ git checkout pull/3615`

Update a local copy of the PR: \
`$ git checkout pull/3615` \
`$ git pull https://git.openjdk.java.net/jdk pull/3615/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3615`

View PR using the GUI difftool: \
`$ git pr show -t 3615`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3615.diff">https://git.openjdk.java.net/jdk/pull/3615.diff</a>

</details>

kelthuzadx commented:
/label add hotspot-compiler
/label add nio

kelthuzadx commented:
/label add core-libs

kelthuzadx commented:
Can I have a review of this PR? Thanks!

AlanBateman commented on src/java.base/share/classes/java/nio/Buffer.java:
```diff
@@ -736,11 +736,8 @@ final int nextPutIndex(int nb) {                    // package-private
-        if ((i < 0) || (i >= limit))
-            throw new IndexOutOfBoundsException();
-        return i;
+        return Objects.checkIndex(i, limit);
```
Changing Buffer.checkIndex to use Objects.checkIndex is okay.

vnkozlov commented:
@veresov please look on C1 changes.

veresov commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -2041,6 +2041,15 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
+          inline_target->name() == ciSymbols::checkIndex_name()) {
+        return;
+      }
+
```
Do we need to keep this flag?

kelthuzadx replied:
> Do we need to keep this flag?

Exactly, the flag should be removed.

Thanks,
Yang

veresov reviewed (commented):
The C1 changes look

dfuch commented on test/hotspot/jtreg/compiler/c1/TestCheckIndexC1Intrinsic.java:
```diff
@@ -0,0 +1,161 @@
+    static void check0() {
+        long res = Preconditions.checkIndex(0, 1, null);
+        Asserts.assertEquals((int)res, 0);
+        try {
+            Preconditions.checkIndex(1, 1, null);
+        } catch (IndexOutOfBoundsException e) {
+            // got it!
+        }
```
I all places where `Precondition.checkIndex` is expected to throw, an AssertionError should  be generated if it doesn't throw:

```
        try {
            Preconditions.checkIndex(1, 1, null);
            throw new AssertionError("Expected IndexOutOfBoundsException not thrown");
        } catch (IndexOutOfBoundsException e) {
            // got it!
        }
```

kelthuzadx replied:
Yes, it does make sense

dfuch commented on test/hotspot/jtreg/compiler/c1/TestCheckIndexC1Intrinsic.java:
```diff
@@ -78,7 +82,8 @@ static void check0() {
             // read fields
-            Preconditions.checkIndex(limit + 1, limit, (s, integers) -> new MyException("Reason:" + s +"::" + integers));
+            Preconditions.checkIndex(limit + 1, limit, (s, integers) -> new MyException("Reason:" + s + "::" + integers));
+            throw new AssertionError("Expected IndexOutOfBoundsException not thrown");
```
nit: well maybe here it should be:
```
throw new AssertionError("Expected MyException not thrown");
```

dfuch commented on test/hotspot/jtreg/compiler/c1/TestCheckIndexC1Intrinsic.java:
```diff
@@ -0,0 +1,166 @@
+
+    static void check1(int i) {
+        try {
+            Preconditions.checkIndex(i, 9999, (s, integers) -> new RuntimeException("ex"));
```
I believe 
```
throw new AssertionError("Expected IndexOutOfBoundsException not thrown");
```

should be added in `check1`...`check4` as well...

kelthuzadx replied:
Hmm.. check1 and check2 should add. check3 and check4 are not necessary since the desired exception is thrown only if i==9999.

kelthuzadx replied:
Hmm.. They would throw desired exceptions only if i==9999.

dfuch reviewed (approved):
Thanks for taking this feedback into account and updating the test!
Note: I only reviewed the test.

PaulSandoz commented:
It was my hope this would eventually happen when we added `Objects.checkIndex` and the underlying intrinsic. Very good!

kelthuzadx commented:
> It was my hope this would eventually happen when we added `Objects.checkIndex` and the underlying intrinsic. Very good!

Hi Paul,

Thank you for noticing this PR.
> It was my hope this would eventually happen when we added `Objects.checkIndex` and the underlying intrinsic.
Yes, this patch adds C1 intrinsic supports for checkIndex, I will replace all variants of checkIndex with Objects.checkIndex in follow-up PRs.

It seems that Object.checkIndex can not meet our needs because it implicitly passes null to Preconditions.checkIndex, but we want to customize exception messages, so we might add extra APIs in Objects while doing the replacement.

> Very good!
Thank you Paul~

Best Regards,
Yang

kelthuzadx commented:
> Looks like now the test fails in the pre-submit tests?

Hi Igor,

Can you take a look at the latest version? Now it passes all pre-submit tests.

Best Regards,
Yang

PaulSandoz commented:
> It seems that Object.checkIndex can not meet our needs because it implicitly passes null to Preconditions.checkIndex, but we want to customize exception messages, so we might add extra APIs in Objects while doing the replacement.
>

It might be possible to directly use `Preconditions.checkIndex` for such purposes, it's package is non-exported but public for reuse within `java.base`. It's used in `VarHandle` `byte` array access code, see [here](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template#L117).

I would caution against adding a public API to support this. The work in  `Preconditions` was the basis for a public API, but it proved complicated. I am glad we did not expose that, it would of made exposing the long accepting `Objects.checkIndex` methods much difficult.

Paul.

kelthuzadx commented:
Hi @veresov, may I ask your help to review this patch? Thanks a lot.

kelthuzadx commented:
Thank you @veresov! 

I'm glad to have more comments from hotspot-compiler group.

Later, I'd like to integrate it if there are no more comments/objections.

Thanks!
Yang

kelthuzadx commented:
/integrate

veresov commented:
@kelthuzadx  Sorry about the delay. Could you please rebase this to the current master and I'll push it. Thanks!

kelthuzadx commented:
@veresov I've rebased to the latest commit and resolved all conflicts. Please take a look at this. Thank you!

veresov commented:
`test/jdk/java/util/Objects/CheckIndex.java` and `test/jdk/java/util/Objects/CheckLongIndex.java` started failing. Please take a look.

```
test CheckIndex.testCheckIndex(0, -2147483647, false): failure
java.lang.AssertionError: Index 0 is out of bounds of [0, -2147483647), but was reported to be within bounds
	at org.testng.Assert.fail(Assert.java:94)
	at CheckIndex.lambda$testCheckIndex$3(CheckIndex.java:103)
	at CheckIndex.testCheckIndex(CheckIndex.java:117)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
test CheckIndex.testCheckIndex(0, -2147483648, false): failure
java.lang.AssertionError: Index 0 is out of bounds of [0, -2147483648), but was reported to be within bounds
	at org.testng.Assert.fail(Assert.java:94)
	at CheckIndex.lambda$testCheckIndex$3(CheckIndex.java:103)
	at CheckIndex.testCheckIndex(CheckIndex.java:120)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
test CheckIndex.testCheckIndex(1, -1, false): failure
java.lang.AssertionError: Index 1 is out of bounds of [0, -1), but was reported to be within bounds
	at org.testng.Assert.fail(Assert.java:94)
	at CheckIndex.lambda$testCheckIndex$3(CheckIndex.java:103)
	at CheckIndex.testCheckIndex(CheckIndex.java:130)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

veresov commented:
Alright, tests pass now. I think we're good to go.

veresov commented:
/sponsor

veresov commented:
I guess you need to do the "integrate" command again.

kelthuzadx commented:
> I guess you need to do the "integrate" command again.

Okay，thank you all for taking time to look at this

/integrate

simonis commented:
This change removed a product flag so I wonder how it could be integrated without a CSR?

tstuefe commented:
> This change removed a product flag so I wonder how it could be integrated without a CSR?

And if the intention was to remove it, should it not have been marked as obsolete first?

kelthuzadx commented:
> This change removed a product flag so I wonder how it could be integrated without a CSR?

It's a diagnostic product flag, so it’ okay to remove it without issuing CSR. But I am not 100% sure.

@dholmes-ora, do you have any comment about this? Thanks!

simonis commented:
Looks like you are right Yi:

https://wiki.openjdk.java.net/display/HotSpot/Hotspot+Command-line+Flags%3A+Kinds%2C+Lifecycle+and+the+CSR+Process

Seems like for diagnostic flags the creation of a CSR is up to the
developer.

Sorry for the confusion.

Yi Yang ***@***.***> schrieb am Sa., 12. Juni 2021, 08:54:

> This change removed a product flag so I wonder how it could be integrated
> without a CSR?
>
> It's a diagnostic product flag, so it’ okay to remove it without issuing
> CSR. But I am not 100% sure.
>
> @dholmes-ora <https://github.com/dholmes-ora>, do you have any comment
> about this? Thanks!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/3615#issuecomment-860010589>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYH5VPC6ZMY76YNJ4C4S4DTSMABPANCNFSM43L2BATA>
> .
>

tstuefe commented:
Hi Yi,

you may need to add the option to the obsolete-flags-table though as described in arguments.cpp:

https://github.com/openjdk/jdk/blob/5cee23a9ed0b7fe2657be7492d9c1f78fcd02ebf/src/hotspot/share/runtime/arguments.cpp#L489-L490

I think the point is to give a customer a grace period where the option is still accepted on the command line. I am not sure if that step is optional though, if one is reasonably sure that the option is unused. Maybe @dholmes-ora can chime in.

Cheers, Thomas

kelthuzadx commented:
> Hi Yi,
> 
> you may need to add the option to the obsolete-flags-table though as described in arguments.cpp:
> 
> https://github.com/openjdk/jdk/blob/5cee23a9ed0b7fe2657be7492d9c1f78fcd02ebf/src/hotspot/share/runtime/arguments.cpp#L489-L490
> 
> I think the point is to give a customer a grace period where the option is still accepted on the command line. I am not sure if that step is optional though, if one is reasonably sure that the option is unused. Maybe @dholmes-ora can chime in.
> 
> Cheers, Thomas

Hi Thomas,

I think what you said is right. It does not take too much time to do this but it can give users a smooth transition for unavailable options! I will create a new PR to do this stuff if there are no objections.

Thanks,
Yang

veresov commented:
It's up to you, we can of course print a warning that the flag has been removed.  In my opinion, we shouldn't waste time on this, that was an obscure non-product flag.
